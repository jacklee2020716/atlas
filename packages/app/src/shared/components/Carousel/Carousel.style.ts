import { StyleFn, makeStyles } from "../../utils"
import { spacing, breakpoints } from "../../theme"

export type CarouselStyleProps = {}

const container: StyleFn = () => ({
	position: "relative",
	display: "flex"
})
const outerItemsContainer: StyleFn = () => ({
	overflow: "hidden"
})

const innerItemsContainer: StyleFn = () => ({
	display: "flex"
})

const navBase: StyleFn = () => ({
	minWidth: spacing.xxxxl,
	minHeight: spacing.xxxxl,
	width: spacing.xxxxl,
	height: spacing.xxxxl,
	position: "absolute"
})

const navLeft: StyleFn = (styles) => ({
	...styles,
	left: 0,
	top: `calc(50% - ${Math.round((parseInt(spacing.xxxxl) + 1) / 2)}px)`
})

const navRight: StyleFn = (styles) => ({
	...styles,
	right: 0,
	top: `calc(50% - ${Math.round((parseInt(spacing.xxxxl) + 1) / 2)}px)`
})

export const useCSS = (props: CarouselStyleProps) => ({
	container: makeStyles([container])(props),
	outerItemsContainer: makeStyles([outerItemsContainer])(props),
	innerItemsContainer: makeStyles([innerItemsContainer])(props),
	navLeft: makeStyles([navBase, navLeft])(props),
	navRight: makeStyles([navBase, navRight])(props)
})
