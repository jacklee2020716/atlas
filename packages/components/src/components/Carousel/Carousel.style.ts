import { css } from "@emotion/core"
import { StyleFn, makeStyles } from "../../utils"

export type CarouselStyleProps = {
	navTopPosition?: string
}

const wrapper: StyleFn = () => ({
	position: "relative"
})
const container: StyleFn = () => ({
	display: "flex",
	overflow: "hidden",
	padding: "1rem"
})

const item: StyleFn = () => ({
	display: "inline-block"
})
const navLeft: StyleFn = (_, { navTopPosition = 69 }) => ({
	position: "absolute",
	left: 0,
	top: `${navTopPosition}px`
})

const navRight: StyleFn = (_, { navTopPosition = 69 }) => ({
	position: "absolute",
	right: 0,
	top: `${navTopPosition}px`
})

export const useCSS = (props: CarouselStyleProps) => ({
	wrapper: makeStyles([wrapper])(props),
	container: makeStyles([container])(props),
	item: makeStyles([item])(props),
	navLeft: makeStyles([navLeft])(props),
	navRight: makeStyles([navRight])(props)
})
