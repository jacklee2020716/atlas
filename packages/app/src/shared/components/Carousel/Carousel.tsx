import React, { useState } from 'react'
import { SerializedStyles } from '@emotion/core'
import { animated, useSpring } from 'react-spring'
import useResizeObserver from 'use-resize-observer'
import { CarouselStyleProps, useCSS } from './Carousel.style'
import NavButton from '../NavButton'

export type CarouselProps = {
  containerCss?: SerializedStyles
  leftControlCss?: SerializedStyles
  rightControlCss?: SerializedStyles
  disableControls?: boolean
  onScroll?: (direction: 'left' | 'right') => void
} & CarouselStyleProps

const Carousel: React.FC<CarouselProps> = ({
  children,
  containerCss,
  leftControlCss,
  rightControlCss,
  disableControls = false,
  onScroll = () => {},
}) => {
  const [scroll, setScroll] = useSpring(() => ({
    transform: `translateX(0px)`,
  }))
  const [carouselOffset, setCarouselOffset] = useState(0)
  const { width: containerWidth = 0, ref: containerRef } = useResizeObserver<HTMLDivElement>()
  const { width: childrenWidth = 0, ref: childrenContainerRef } = useResizeObserver<HTMLDivElement>()

  const styles = useCSS({})

  const maxScrollOffset = childrenWidth - containerWidth

  const showLeftControl = !disableControls && carouselOffset > 0
  const showRightControl = !disableControls && carouselOffset < maxScrollOffset

  const handleScroll = (direction: 'left' | 'right') => {
    if (containerWidth == null) {
      return
    }
    let scrollAmount
    switch (direction) {
      case 'left': {
        // Prevent overscroll on the left
        const newOffset = carouselOffset - containerWidth
        scrollAmount = newOffset < 0 ? 0 : newOffset
        onScroll('left')
        break
      }
      case 'right': {
        // Prevent overscroll on the right
        const newOffset = carouselOffset + containerWidth
        scrollAmount = newOffset > maxScrollOffset ? maxScrollOffset : newOffset
        onScroll('right')
        break
      }
    }
    setCarouselOffset(scrollAmount)
    setScroll({
      transform: `translateX(-${scrollAmount}px)`,
    })
  }

  if (!Array.isArray(children)) {
    return <>{children}</>
  }

  return (
    <div css={[styles.container, containerCss]}>
      <div css={styles.outerItemsContainer} ref={containerRef}>
        <animated.div css={styles.innerItemsContainer} style={scroll}>
          <div css={styles.innerItemsContainer} ref={childrenContainerRef}>
            {children.map((element, idx) => (
              <React.Fragment key={`Carousel-${idx}`}>{element}</React.Fragment>
            ))}
          </div>
        </animated.div>
      </div>
      {showLeftControl && (
        <NavButton outerCss={[styles.navLeft, leftControlCss]} direction="left" onClick={() => handleScroll('left')} />
      )}
      {showRightControl && (
        <NavButton
          outerCss={[styles.navRight, rightControlCss]}
          direction="right"
          onClick={() => handleScroll('right')}
        />
      )}
    </div>
  )
}

export default Carousel
