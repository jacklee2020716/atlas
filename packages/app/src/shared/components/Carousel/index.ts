import Carousel, { CarouselProps } from './Carousel'
import { CAROUSEL_CONTROL_SIZE } from './Carousel.style'

export { Carousel as default, CAROUSEL_CONTROL_SIZE }
export type { CarouselProps }
