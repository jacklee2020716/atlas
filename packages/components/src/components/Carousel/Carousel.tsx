import React, { useState, useRef, useEffect, useCallback, useMemo } from "react";
import { css, SerializedStyles } from "@emotion/core";
import { animated, useSpring } from "react-spring";
import useResizeObserver from "use-resize-observer";
import { useCSS, CarouselStyleProps } from "./Carousel.style";
import NavButton from "../NavButton";

type CarouselProps = {
	children: React.ReactNode;
	containerCss: SerializedStyles;
	leftControlCss: SerializedStyles;
	rightControlCss: SerializedStyles;
	onScroll: (direction: "left" | "right") => void;
} & CarouselStyleProps;

const Carousel: React.FC<Partial<CarouselProps>> = ({
	children,
	containerCss,
	leftControlCss,
	rightControlCss,
	onScroll = () => {},
}) => {
	if (!Array.isArray(children)) {
		return <>{children}</>;
	}
	let [props, set] = useSpring(() => ({
		transform: `translateX(0px)`,
	}));
	const [x, setX] = useState(0);
	const { width: containerWidth, ref: containerRef } = useResizeObserver<HTMLDivElement>();
	const elementsRefs = useRef<(HTMLDivElement | null)[]>([]);
	const [childrensWidth, setChildrensWidth] = useState(0);
	useEffect(() => {
		elementsRefs.current = elementsRefs.current.slice(0, children.length);
		const childrensWidth = elementsRefs.current.reduce(
			(accWidth, el) => (el != null ? accWidth + el.clientWidth : accWidth),
			0
		);
		setChildrensWidth(childrensWidth);
	}, [children.length]);

	const styles = useCSS({});
	return (
		<div css={[styles.container, containerCss]}>
			<div css={styles.itemsContainer} ref={containerRef}>
				{children.map((element, idx) => (
					<animated.div
						style={props}
						key={`Carousel-${idx}`}
						ref={(el) => {
							elementsRefs.current[idx] = el;
							return el;
						}}
					>
						{element}
					</animated.div>
				))}
			</div>
			<NavButton
				outerCss={[styles.navLeft, leftControlCss]}
				direction="left"
				onClick={() => {
					handleScroll("left");
				}}
			/>
			<NavButton
				outerCss={[styles.navRight, rightControlCss]}
				direction="right"
				onClick={() => {
					handleScroll("right");
				}}
			/>
		</div>
	);

	function handleScroll(direction: "left" | "right") {
		if (containerWidth == null) {
			return;
		}
		let scrollAmount;
		switch (direction) {
			case "left": {
				// Prevent overscroll on the left
				scrollAmount = x + containerWidth >= 0 ? 0 : x + containerWidth;
				onScroll("left");
				break;
			}
			case "right": {
				// Prevent overscroll on the right
				scrollAmount =
					x - containerWidth <= -(childrensWidth - containerWidth)
						? -(childrensWidth - containerWidth)
						: x - containerWidth;
				onScroll("right");
				break;
			}
		}
		setX(scrollAmount);
		set({
			transform: `translateX(${scrollAmount}px)`,
		});
	}
};

export default Carousel;
