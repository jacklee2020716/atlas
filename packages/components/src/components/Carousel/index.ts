import { memo } from "react"
import Carousel from "./Carousel"
export default memo(Carousel)
