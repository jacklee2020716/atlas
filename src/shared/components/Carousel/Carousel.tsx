import React, { useState, useLayoutEffect, useRef } from 'react'
import { GliderProps, GliderMethods } from 'react-glider'

import { Container, StyledGlider, Arrow } from './Carousel.style'

import 'glider-js/glider.min.css'

type CarouselProps = {
  trackPadding?: string
} & GliderProps

type TrackProps = {
  className?: string
  padding?: string
}
const Track: React.FC<TrackProps> = ({ className = '', ...props }) => (
  <div className={`glider-track ${className}`} {...props} />
)

const RightArrow = <Arrow name="chevron-right" />
const LeftArrow = <Arrow name="chevron-left" />

const Carousel: React.FC<CarouselProps> = ({
  children,
  trackPadding = '0',
  className,
  itemWidth = 300,
  slidesToShow = 'auto',
  ...gliderProps
}) => {
  //  The GliderMethods type only has methods and I need the full instance
  const gliderRef = useRef<GliderMethods & { ele: HTMLDivElement }>()
  const [arrows, setArrows] = useState<{ prev: HTMLButtonElement; next: HTMLButtonElement } | undefined>(undefined)

  useLayoutEffect(() => {
    if (gliderRef.current) {
      const glider = gliderRef.current.ele
      const prevArrow = glider.previousSibling as HTMLButtonElement
      const nextArrow = glider.nextSibling as HTMLButtonElement

      setArrows({ prev: prevArrow, next: nextArrow })
    }
  }, [])

  // This is needed because react-glider will render arrows only if the arrows option is undefined, so arrows won't display if you pass an object to StyledGlider
  React.useLayoutEffect(() => {
    if (gliderRef.current && arrows) {
      const { prev: prevArrow, next: nextArrow } = arrows
      const container = gliderRef.current.ele.parentElement
      if (container) {
        container.insertBefore(prevArrow, gliderRef.current.ele)
        container.appendChild(nextArrow)
      }
    }
  }, [arrows])

  return (
    <Container trackPadding={trackPadding} className={className}>
      <StyledGlider
        addTrack
        skipTrack
        hasArrows
        draggable
        ref={gliderRef as React.RefObject<GliderMethods>}
        iconLeft={LeftArrow}
        iconRight={RightArrow}
        itemWidth={itemWidth}
        slidesToShow={slidesToShow}
        // Akward conversion needed until this is resolved: https://github.com/hipstersmoothie/react-glider/issues/36
        arrows={(arrows as unknown) as { prev: string; next: string }}
        {...gliderProps}
      >
        <Track padding={trackPadding}>{children}</Track>
      </StyledGlider>
    </Container>
  )
}
export default Carousel
