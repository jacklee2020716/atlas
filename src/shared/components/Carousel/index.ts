import type { CarouselProps } from './Carousel'
import Carousel from './Carousel'

export default Carousel
export type { CarouselProps }
