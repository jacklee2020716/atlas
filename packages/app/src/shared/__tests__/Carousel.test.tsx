import React from 'react'
import { shallow } from 'enzyme'
import Carousel from '@/shared/components/Carousel'
describe('Carousel component', () => {
  it('Should render Carousel correctly', () => {
    expect(
      shallow(
        <Carousel>
          <div>Carousel child 1</div>
          <div>Carousel child 1</div>
          <div>Carousel child 1</div>
        </Carousel>
      )
    ).toBeDefined()
  })
})
