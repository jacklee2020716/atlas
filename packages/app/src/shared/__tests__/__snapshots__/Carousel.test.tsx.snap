// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel component Should render Carousel correctly 1`] = `ShallowWrapper {}`;
